libdistopia/src/distopia.cpp

```diff
@@ -57,20 +58,91 @@ namespace distopia {
                 vz = lz * dsz;
             };
         };
-        template <class D, typename T = hn::TFromD<D>>
+
+        template <class D, typename T = hn::TFromD<D>, typename V = hn::VFromD<D>>
         struct TriclinicBox {
             hn::VFromD<D> xx, xy, yy, xz, yz, zz;
+            hn::VFromD<D> ix, iy, iz;
 
             explicit TriclinicBox(D d, const T *sbox) {
                 this->xx = hn::Set(d, sbox[0]);
                 this->xy = hn::Set(d, sbox[1]); this->yy = hn::Set(d, sbox[2]);
                 this->xz = hn::Set(d, sbox[3]); this->yz = hn::Set(d, sbox[4]); this->zz = hn::Set(d, sbox[5]);
+                // inverse of diagonal
+                this->ix = hn::Set(d, 1/sbox[0]);
+                this->iy = hn::Set(d, 1/sbox[2]);
+                this->iz = hn::Set(d, 1/sbox[5]);
             };
 
-            void MinimiseVectors(hn::VFromD<D> &vx,
-                                 hn::VFromD<D> &vy,
-                                 hn::VFromD<D> &vz) const {
+            void ShiftIntoPrimaryUnitCell(V &vx, V &vy, V &vz) const {
+                auto sz = hn::Floor(this->iz * vz);
+                vz += sz * this->zz;
+                vy += sz * this->yz;
+                vx += sz * this->xz;
+                auto sy = hn::Floor(this->iy * vy);
+                vy += sy * this->yy;
+                vx += sy * this->xy;
+                auto sx = hn::Floor(this->ix, vx);
+                vx += sx * this->xx;
+            }
 
+            void MinimiseVectors(V &vx,
+                                 V &vy,
+                                 V &vz) const {
+                // check all 27 (3x3x3) possibilities of adding/subtracting box vectors and choose minimum
+                V vmin[3];
+                hn::ScalableTag<T> d;
+                auto dsq_min = hn::Set(d, HUGE_VALF);
+
+                for (int i = -1; i < 2; ++i) {  // loop over subtracting, keeping or adding X dimension from vector
```

Comment:
Could we make these loop counters `size_t`? Is there any reason they have to go from  -1, 0, 1 and cant go  0, 1, 3 ?

size_t can help with the loop vectorisation.